sqld/src/database/libsql.rs

```diff
@@ -266,11 +185,153 @@ impl LibSqlDb {
     }
 }
 
+struct Connection {
+    state: ConnectionState,
+    conn: rusqlite::Connection,
+    timed_out: bool,
+}
+
+impl Connection {
+    fn new(
+        path: &Path,
+        wal_hook: impl WalHook + Send + Clone + 'static,
+        with_bottomless: bool,
+    ) -> anyhow::Result<Self> {
+        Ok(Self {
+            conn: open_db(path, wal_hook, with_bottomless)?,
+            state: ConnectionState::initial(),
+            timed_out: false,
+        })
+    }
+
+    fn run(&mut self, pgm: Program) -> Vec<Option<QueryResult>> {
+        let mut results = Vec::with_capacity(pgm.steps.len());
+
+        for step in pgm.steps {
+            let res = self.execute_step(&step, &results);
+            results.push(res);
+        }
+
+        results
+    }
+
+    fn execute_step(
+        &mut self,
+        step: &Step,
+        results: &[Option<QueryResult>],
+    ) -> Option<QueryResult> {
+        let enabled = match step.cond.as_ref() {
+            Some(cond) => match eval_cond(cond, results) {
+                Ok(enabled) => enabled,
+                Err(e) => return Some(Err(e)),
+            },
+            None => true,
+        };
+
+        enabled.then(|| self.execute_query(&step.query))
```

Comment:
Nice! I didn't know about `bool::then()`, but I often thought that it would be useful :)